src/core/datastore/sync/sync-manager.js

```diff
@@ -347,4 +352,66 @@ export class SyncManager {
         return Promise.reject(new SyncError('Invalid sync event name'));
     }
   }
+
+  _getPullLimit(pullOptions) {
+    const limitOption = pullOptions && pullOptions.pullLimit;
+    return limitOption || Infinity;
+  }
+
+  _getInternalPullQuery({ filter, sort, fields }, pullLimit) {
+    const query = new Query({ filter, sort, fields });
+    query.skip = 0;
+    query.limit = null; // the current default value
+
+    if (pullLimit && pullLimit < Infinity) {
+      query.limit = pullLimit;
+    }
+
+    return query;
+  }
+
+  _getPaginationQueries({ filter, sort, fields }, pageSize = maxEntityLimit, totalCount) {
+    const queryCount = Math.ceil(totalCount / pageSize);
+    const queries = times(queryCount, (i) => {
+      const query = new Query({ filter, sort, fields });
+      query.skip = i * pageSize;
+      query.limit = Math.min(totalCount - query.skip, pageSize);
+      return query;
+    });
+    return queries;
+  }
+
+  _fetchAndUpdateEntities(collection, query) {
+    return this._networkRepo.read(collection, query)
+      .then((entities) => {
+        if (isEmpty(entities)) {
+          return Promise.resolve();
+        }
+        return this._getOfflineRepo()
+          .then(repo => repo.update(collection, entities));
+      });
+  }
+
+  _executePaginationQueries(collection, deleteQuery, queries) {
+    return this._getOfflineRepo()
+      .then(repo => repo.delete(collection, deleteQuery))
+      .then(() => {
+        // TODO: throttle max request count - reuse from push?
+        return forEachAsync(queries, (query) => {
+          return this._fetchAndUpdateEntities(collection, query);
+        });
+      });
+  }
+
+  _paginate(collection, userQuery, options) {
+    const query = new Query({ filter: userQuery.filter }); // ignore sort, it's irrelevant
+    return this._networkRepo.count(collection, query)
+      .then((totalCount) => {
+        const pullLimit = this._getPullLimit(options);
```

Comment:
I haven't understood this parameter. What's `pullLimit` supposed to signify? Typically, developers will limit their results by providing a query, not a numerical limit.

Reply:
I mentioned it in the PR "opening" comment. It's a fix for the scenario we discussed after standup yesterday and I mentioned in the doc. It allows you to do an auto paginated `sort({'_kmd.ect': 1}).limit(21000)`. You reminded me they can still do it by calling pull() with no auto pagination a few times (for when they want >10k results). But I already had it in there. And I feel it doesn't do harm, but I will remove it if it doesn't seem useful and/or is confusing.